lib/web_ui/lib/src/engine/host_node.dart

```diff
@@ -0,0 +1,183 @@
+  ///
+  /// If the node already exists in this document, it will be removed from its
+  /// current parent node, then added to this node.
+  ///
+  /// This method is more efficient than `nodes.add`, and is the preferred
+  /// way of appending a child node.
+  ///
+  /// See:
+  /// * [Node.appendChild](https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild)
+  html.Node append(html.Node node);
+
+  /// Returns true if this node contains the specified node.
+  /// See:
+  /// * [Node.contains](https://developer.mozilla.org/en-US/docs/Web/API/Node.contains)
+  bool contains(html.Node? other);
+
+  /// Returns the currently wrapped [html.Node].
+  html.Node get node;
+
+  /// A modifiable list of this node's children.
+  List<html.Node> get nodes;
+
+  /// Finds the first descendant element of this document that matches the
+  /// specified group of selectors.
+  ///
+  /// [selectors] should be a string using CSS selector syntax.
+  ///
+  /// ```dart
+  /// var element1 = document.querySelector('.className');
+  /// var element2 = document.querySelector('#id');
+  /// ```
+  ///
+  /// For details about CSS selector syntax, see the
+  /// [CSS selector specification](http://www.w3.org/TR/css3-selectors/).
+  ///
+  /// See:
+  /// * [Document.querySelector](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector)
+  html.Element? querySelector(String selectors);
+
+  /// Finds all descendant elements of this document that match the specified
+  /// group of selectors.
+  ///
+  /// [selectors] should be a string using CSS selector syntax.
+  ///
+  /// ```dart
+  /// var items = document.querySelectorAll('.itemClassName');
+  /// ```
+  ///
+  /// For details about CSS selector syntax, see the
+  /// [CSS selector specification](http://www.w3.org/TR/css3-selectors/).
+  ///
+  /// See:
+  /// * [Document.querySelectorAll](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelectorAll)
+  List<html.Node> querySelectorAll(String selectors);
+}
+
+/// A [HostNode] implementation, backed by a [html.ShadowRoot].
+///
+/// This is the preferred flutter implementation, but it might not be supported
+/// by all browsers yet.
+///
+/// The constructor might throw when calling `attachShadow`, if ShadowDOM is not
+/// supported in the current environment. In this case, a fallback [ElementHostNode]
+/// should be created instead.
+class ShadowDomHostNode implements HostNode {
+  late html.ShadowRoot _shadow;
+
+  /// Build a HostNode by attaching a [html.ShadowRoot] to the `root` element.
+  ///
+  /// This also calls [applyGlobalCssRulesToSheet], defined in dom_renderer.
+  ShadowDomHostNode(html.Element root) {
+    _shadow = root.attachShadow(<String, String>{
+      'mode': 'open',
+      'delegatesFocus': 'true',
+    });
+
+    final html.StyleElement shadowRootStyleElement = html.StyleElement();
+    // The shadowRootStyleElement must be appended to the DOM, or its `sheet` will be null later.
```

Comment:
Can we assert that `root` is appended to the DOM? `assert(root.isConnected);`?

Reply:
`isConnected` is `bool?`... Should we skip the assert if we can't determine what's its value?

Reply:
`isConnected` SGTM.